### DISCUSSION TITLE
Implementation Discussion for Message Filtering

### DISCUSSION THREAD

**Discussion by 96-LB** [Tyrrrz/DiscordChatExporter#591] [General]
I'm opening this discussion thread to explore possible approaches to implementing a solution for #160. The ability to filter which messages are exported and downloaded is perhaps one of the most requested features.

Relevant issues include: #28 #178 #219 #336 #343 #361 #384 #430 #566
***
As the scope of these requests varies from "export the messages of a specific user" to "compound filters with boolean operators", we should probably first determine to what extent the feature would encompass. I'm personally in favor of adding a query parameter which uses syntax similar to the search bar in the Discord client (e.g. `from:236257776421175296 before:2021`), but as far as I know, Discord doesn't expose their search function in their API. Discord's search is also complex (e.g. supports spaces like `from:96 LB#7383`), quirky (e.g. searching `ring` brings up results for `ringing` but not `r`), and limited (no support for "OR" or "NOT" queries). As such, implementing a close copy would be basically impossible.

I think we could reasonably implement a simple `key:value` query syntax. I did an extremely cursory search for useful libraries and nothing stood out, but I'm sure something like that exists. It probably wouldn't be difficult to build on our own either; though it may be a little more of a challenge to search for text the way Discord does unless we add our own properties such as `contains:` or `startswith:` (which I don't think is a bad idea).

There's a lot to discuss with this feature, so hopefully this gets the ball rolling.

**Comment by Tyrrrz** [comment]
Hello!

> I'm personally in favor of adding a query parameter which uses syntax similar to the search bar in the Discord client

Yes, I think it's the way to go. This means we won't have to come up with something entirely new ourselves and most users will be familiar with the syntax.

> Discord doesn't expose their search function in their API. Discord's search is also complex (e.g. supports spaces like from:96 LB#7383), quirky (e.g. searching ring brings up results for ringing but not r), and limited (no support for "OR" or "NOT" queries).

1. I believe the `from:96 LB#7383` thing works when you type and select the user from a dropdown. I think in such case Discord treats the object you selected as a token and not just as a string of text, so it disregards the spaces. In our case we won't have a dropdown, so escaping spaces with double quotes can be a reasonable workaround.
2. Fuzzy search (i.e. `ringing`/`ring`) can definitely be a stretch goal. We don't have to make a 100% copy of Discord's search query syntax, but the closer we can get, the better.

> I think we could reasonably implement a simple key:value query syntax. I did an extremely cursory search for useful libraries and nothing stood out, but I'm sure something like that exists. It probably wouldn't be difficult to build on our own either; though it may be a little more of a challenge to search for text the way Discord does unless we add our own properties such as contains: or `startswith: (which I don't think is a bad idea).

From the outset it looks like even simple regex expressions can do because there aren't any nested tokens, so a recursive-descent parser (like the one I built for markdown) is not necessary.

**Reply by 96-LB** [reply]
> I believe the `from:96 LB#7383` thing works when you type and select the user from a dropdown. I think in such case Discord treats the object you selected as a token and not just as a string of text, so it disregards the spaces. In our case we won't have a dropdown, so escaping spaces with double quotes can be a reasonable workaround.

Discord actually tokenizes usernames with spaces even if you manually type them in, as long as the discriminator is included (e.g. `96 LB` does not work). I think, for the `from:` field in specific, Discord looks either for a user ID or for some string ending in a discriminator. However, I'm also in favor of using double quotes to group strings with spaces, as that would keep options open for other parameters.

> Fuzzy search (i.e. ringing/ring) can definitely be a stretch goal. We don't have to make a 100% copy of Discord's search query syntax, but the closer we can get, the better.

From my observations, Discord's fuzzy search seems to deal primarily with common suffixes such as "-s", "-d", "-er", "-ing", and "-ed". A notable exception is "-ly". In a few cases (I think mostly with "-s") you can also stack suffxies, such as searching for `youings` with the query `you` or vice versa.

> From the outset it looks like even simple regex expressions can do because there aren't any nested tokens, so a recursive-descent parser (like the one I built for markdown) is not necessary.

Agreed. I haven't looked too deeply into it yet, but would your `RegexMatcher` class (and/or the other classes within the Matching folder) be suitable to be reused for this?

**Reply by Tyrrrz** [reply]
> Discord actually tokenizes usernames with spaces even if you manually type them in, as long as the discriminator is included (e.g. 96 LB does not work). I think, for the from: field in specific, Discord looks either for a user ID or for some string ending in a discriminator. However, I'm also in favor of using double quotes to group strings with spaces, as that would keep options open for other parameters.

:+1:

> From my observations, Discord's fuzzy search seems to deal primarily with common suffixes such as "-s", "-d", "-er", "-ing", and "-ed". A notable exception is "-ly". In a few cases (I think mostly with "-s") you can also stack suffxies, such as searching for youings with the query you or vice versa.

Let's worry about it in the future ;)

> Agreed. I haven't looked too deeply into it yet, but would your RegexMatcher class (and/or the other classes within the Matching folder) be suitable to be reused for this?

Hmm, I think they would be overkill because we really only have a single pattern: `<filter>:<value>`. In markdown we have a myriad of completely different patterns, so those matchers were used as a way properly apply them in sequence (e.g. if match2 triggers earlier than match1 then it wins). From what I can tell, the search query syntax won't have these challenges.

**Reply by 96-LB** [reply]
I see. What do you think the best course of action is for handling text content queries (i.e. searching for words within the messages)? I'd imagine we could simply treat every token that doesn't match the `<filter>:<value>` pattern as text and go from there, unless we go the route of adding our own filters as I previously spitballed (`startswith:`, `contains:`, etc.).

If we do the former in order to more closely mimic Discord, I'm not entirely sure what the best way to actually perform the matching is, as I can't concretely identify Discord's behavior. In the past, I believe queries like `foo baz` would match `foo bar baz` but not `baz foo`, though I also remember it behaving differently at times. Right now it seems to match any message in which `foo` and `baz` are within a few words of each other in any order, but the specific search function seems to change arbitrarily. If we consider the use case of word-searching for DCE, I'd wager that it's probably better to have a simple but well-defined behavior—such as simply checking if each text token is present anywhere in the message—instead of trying to replicate Discord's methods. Thoughts?

**Reply by Tyrrrz** [reply]
> I see. What do you think the best course of action is for handling text content queries (i.e. searching for words within the messages)? I'd imagine we could simply treat every token that doesn't match the `<filter>:<value>` pattern as text and go from there, unless we go the route of adding our own filters as I previously spitballed (`startswith:`, `contains:`, etc.).

Let's treat every non key-value token as text and then basically apply it with `Contains(...)` function. I'm not sure if filters like `startswith:` etc. are even useful in practice, so let's wait before introducing them.

> If we do the former in order to more closely mimic Discord, I'm not entirely sure what the best way to actually perform the matching is, as I can't concretely identify Discord's behavior. In the past, I believe queries like `foo baz` would match `foo bar baz` but not `baz foo`, though I also remember it behaving differently at times. Right now it seems to match any message in which `foo` and `baz` are within a few words of each other in any order, but the specific search function seems to change arbitrarily. If we consider the use case of word-searching for DCE, I'd wager that it's probably better to have a simple but well-defined behavior—such as simply checking if each text token is present anywhere in the message—instead of trying to replicate Discord's methods. Thoughts?

This is an interesting point because there's a big difference in how those filters are used in Discord vs. how they will be used here. In Discord, you search for messages usually to find one or maybe a few very specific messages. So the way the search works is optimized to return as few results as possible, ordered by relevancy. In our case, we will filter which messages we will export, so we will still likely end up with a lot of results. Also, we won't be re-ordering them (so we won't, for example, push more relevant matches higher). I imagine people that will use this feature in DCE will mostly use filters like `has:attachments` to get all messages with images, or `user:Foo#1234` to get all messages from a specific user (for this we may need to support multiple `user:xyz` tokens by the way). I doubt a lot of people will be searching for text. We shouldn't try to match Discord's search functionality too much because fundamentally it will be used in a completely different way here. Using the same syntax for `has:`/`user:` etc. filters is a benefit though and I think it applies well for us too.

**Reply by 96-LB** [reply]
> (for this we may need to support multiple `user:xyz` tokens by the way)

You raise a good point here concerning the use of multiple tokens. I agree that it makes sense to treat `user:foo user:bar` as `user:foo OR user:bar`, since obviously no message can satisfy both conditions. At the same time, however, the same logic probably shouldn't apply to something like `user:foo has:attachment`, where the user would probably except to find only messages sent by `foo` with an attachment.

Probably the simplest way of dealing with this would be to check whether each message satisfies at least one of every "type" of token if present (e.g. <any `user:` query is met> AND <any `has:` query is met>, etc.). As the only built-in Discord search tokens that we need to worry about are `user:`, `has:`, and `mentions:` (as `in:`, `before:`, `during:`/`on:`, and `after:` are redundant), that behaviour seems most intuitive to me. The only hitch I can see is applying this to text tokens where, for example, a search like `living room` could return results for all sorts of different rooms—but even this could probably be sidestepped with the use of double quotes (i.e. `"living room"`), so I feel like this is the way to go.

Someone in the main issue suggested implementing logical operators, but that seems unnecessarily complex to me as typically only one combination makes sense. I suppose it's possible that negation is useful?

**Reply by Tyrrrz** [reply]
Yeah, negation could be useful and potentially `AND`/`OR` too. For example, when combining `has:attachment` and `has:link`, I can imagine there could be cases where you'd want AND and also cases where you'd want OR. That could also be for later though, and for now we can just make sensible defaults that satisfy 95% use cases. I think it makes sense to have the `OR` combinator applied by default to both `user` and `has` for the sake of consistency, later on we can think about making an explicit `AND` as well. For negation operator, we can go with GitHub's syntax and use the `-` symbol, i.e. `-has:attachment` = excludes messages with attachments.

As for text, I think what you suggested makes sense, including the double quotes to treat multiple words as a single token.

**Reply by 96-LB** [reply]
> Yeah, negation could be useful and potentially AND/OR too. For example, when combining `has:attachment` and `has:link`, I can imagine there could be cases where you'd want AND and also cases where you'd want OR. That could also be for later though, and for now we can just make sensible defaults that satisfy 95% use cases.

Ah, I see. 👍

> For negation operator, we can go with GitHub's syntax and use the - symbol, i.e. -has:attachment = excludes messages with attachments.

I agree with this. If we build this in, we should apply an implicit AND/NOR across negations, right? Interpreting `-user:foo -user:bar` as `-user:foo OR -user:bar` would match everything and not be very useful. Treating a mix of positives and negations seems a bit more tricky, as `has:attachment -has:link` could be interpreted as either "all attachments without links" or "nothing with a link unless it has an attachment". The only way around that I see besides explicit logical operators is to make the query order-sensitive (e.g. `has:attachment -has:link` is the former behavior while `-has:link has:attachment` is the latter or vice versa), unless you have a better idea. (Does GitHub handle this case? I haven't really used the search function much.)

**Reply by Tyrrrz** [reply]
> I agree with this. If we build this in, we should apply an implicit AND/NOR across negations, right? Interpreting `-user:foo -user:bar` as `-user:foo OR -user:bar` would match everything and not be very useful. Treating a mix of positives and negations seems a bit more tricky, as `has:attachment -has:link` could be interpreted as either "all attachments without links" or "nothing with a link unless it has an attachment". The only way around that I see besides explicit logical operators is to make the query order-sensitive (e.g. `has:attachment -has:link` is the former behavior while `-has:link has:attachment` is the latter or vice versa), unless you have a better idea. (Does GitHub handle this case? I haven't really used the search function much.)

I don't think using order for resolution is a good idea. I'd imagine most people are used to order being irrelevant in such scenarios. GitHub unfortunately has a different use case as well, in which the goal of the search is to narrow down the results, meaning that all operands are effectively combined with `AND`.

On a second thought, maybe it's better to have `AND` the default operator and make `OR` explicit even in cases where it's the only thing that makes sense, i.e.:

```
has:attachment -has:link (user:Tyrrrz | user:96-LB)
```

(using `|` as `OR` to avoid ambiguity as to whether the 'or' is a text token or an operator)

The above query should return all messages that satisfy all following conditions:
- Have an attachment
- Do not have a link
- Are sent either by Tyrrrz or 96-LB

**Reply by 96-LB** [reply]
> I don't think using order for resolution is a good idea. I'd imagine most people are used to order being irrelevant in such scenarios.

Yeah, I don't believe it's the best solution either; I just couldn't think of any other way to resolve that case without an explicit operator.

> On a second thought, maybe it's better to have AND the default operator and make OR explicit even in cases where it's the only thing that makes sense, i.e.:
> ```
> has:attachment -has:link (user:Tyrrrz | user:96-LB)
> ```

I'm in support of this syntax. We *are* getting pretty close to a recursive parser at this point; should we support things like `(user:a (mentions:b | mentions:c)) | ((user:b | user:c) mentions:a)`? It seems to me that we'd be one step away and could simply recursively parse any query within parentheses.

Another thing to consider is that usernames can contain `(`, `|`, and `)`, so we'd likely need a way to escape those characters. I suppose we could ignore any instances of them within double quotes since we're already using them to escape spaces.

Also, it might be worth implementing `before:`/`after:`/`during:` now, since they could potentially have a place within complex queries (e.g. exporting the messages of a gameshow host whose identity switched between seasons could be done as `(during:2020 user:a) | (during:2021 user:b)`).

**Reply by Tyrrrz** [reply]
Yeah at this point it actually is recursive. I still wouldn't use the existing abstractions we have in `IMatcher` and co because they are very specialized for dealing with Discord's markdown syntax which, although recursive, is not balanced in the same way a normal context-free language would be (even unlike normal markdown; Discord's markdown is very special). So instead I'd look into [Pidgin](https://github.com/benjamin-hodgson/Pidgin) or [Superpower](https://github.com/datalust/superpower). I'm more familiar with the latter but as far as I know Pidgin is the fastest.

Let's hold with before/after/during simply because it causes a bit of confusion with the existing export date ranges that we have. If someone asks for that feature, we probably will need to think on how to converge them.

**Reply by 96-LB** [reply]
> So instead I'd look into [Pidgin](https://github.com/benjamin-hodgson/Pidgin) or [Superpower](https://github.com/datalust/superpower). I'm more familiar with the latter but as far as I know Pidgin is the fastest.

I haven't worked with either library in the past—though I'm both willing and eager to learn!—so it doesn't matter a whole lot to me which one we use. I skimmed the readme's of both and would be fine with either one. Do you have a preference?

> Let's hold with before/after/during simply because it causes a bit of confusion with the existing export date ranges that we have. If someone asks for that feature, we probably will need to think on how to converge them.

👍

**Reply by Tyrrrz** [reply]
> I haven't worked with either library in the past—though I'm both willing and eager to learn!—so it doesn't matter a whole lot to me which one we use. I skimmed the readme's of both and would be fine with either one. Do you have a preference?

Let's go with Superpower. Performance is not a concern here because we'd only be parsing the query once per each export (unlike the markdown parser which runs on each message).

**Reply by 96-LB** [reply]
Sounds good. I'll take some time to experiment with Superpower and then could look into getting something started for the grammar in the near future. I don't have any other concerns at the moment, though I'll probably open a work-in-progress PR at some point moving forward to continue weighing options regarding the actual functional implementation. Is there anything else you wanted to discuss right now?

**Reply by Tyrrrz** [reply]
Sounds good to me :)

**Reply by haiitsstrawberry** [reply]
i think it would also be helpful if you had something like user:bot or user:webhook and the startswith: is useful because it can filter out bot commands(which is what i need in my situation) also for the logical operators i think if it isn't specified you can just take what discord defaults when searched e.g.` strand (user:foo user:bar)` searches the word strand from either user foo or user bar

**Reply by 96-LB** [reply]
Yeah, I'm in support of eventually integrating bot/user/webhook filters and some more advanced text matching at some point in the future. Also, I just realized that Discord actually uses the `from:` keyword instead of `user:` like we've been saying 😅. Should we include an alias for that?

**Reply by haiitsstrawberry** [reply]
Woops, i mixed up `from:` and `user:`, how about `from:` specifies which user(LB, Tyrrrz) and `user:` specifies which kind of user(bot/webhook/user). For aliases idk

**Reply by 96-LB** [reply]
I think `user:` could easily get mixed up with `from:` if we do that. I'd prefer something naming it something like `type:`.

**Reply by haiitsstrawberry** [reply]
sounds good
